VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/General/ImplementEqualsAndGetHashCode/ImplementEqualsAndGetHashCodeCodeFix.cs

```diff
@@ -0,0 +1,256 @@
+
+            context.RegisterCodeFix(
+                CodeAction.Create(string.Format(VSDiagnosticsResources.ImplementEqualsAndGetHashCodeCodeFixTitle),
+                    x => ImplementEqualsAndGetHashCodeAsync(context.Document, root, statement),
+                    ImplementEqualsAndGetHashCodeAnalyzer.Rule.Id),
+                diagnostic);
+        }
+
+        private async Task<Document> ImplementEqualsAndGetHashCodeAsync(Document document, SyntaxNode root, SyntaxNode statement)
+        {
+            var model = await document.GetSemanticModelAsync();
+
+            var objectSymbol = model.Compilation.GetSpecialType(SpecialType.System_Object);
+            var objectEquals = objectSymbol.GetMembers().OfType<IMethodSymbol>()
+                    .First(method => method.MetadataName == nameof(Equals) && method.Parameters.Count() == 1);
+
+            if (statement is ClassDeclarationSyntax)
+            {
+                var classDeclaration = (ClassDeclarationSyntax) statement;
+
+                var classSymbol = model.GetDeclaredSymbol(classDeclaration);
+
+                var newRoot = root.ReplaceNode(classDeclaration,
+                    classDeclaration.AddMembers(GetEqualsMethod(model, classDeclaration.Identifier, classDeclaration.Members, classSymbol, objectEquals),
+                        GetGetHashCodeMethod(model, classDeclaration.Members)));
+                return document.WithSyntaxRoot(newRoot);
+            }
+            else
+            {
+                var structDeclaration = (StructDeclarationSyntax)statement;
+
+                var structSymbol = model.GetDeclaredSymbol(structDeclaration);
+
+                var newRoot = root.ReplaceNode(structDeclaration,
+                    structDeclaration.AddMembers(GetEqualsMethod(model, structDeclaration.Identifier, structDeclaration.Members, structSymbol, objectEquals),
+                        GetGetHashCodeMethod(model, structDeclaration.Members)));
+                return document.WithSyntaxRoot(newRoot);
+            }
+        }
+
+        private MethodDeclarationSyntax GetEqualsMethod(SemanticModel model, SyntaxToken identifier, SyntaxList<SyntaxNode> members, INamedTypeSymbol typeSymbol, IMethodSymbol objectEquals)
+        {
+            var publicModifier = SyntaxFactory.Token(SyntaxKind.PublicKeyword);
+            var overrideModifier = SyntaxFactory.Token(SyntaxKind.OverrideKeyword);
+
+            var parameter = SyntaxFactory.Parameter(SyntaxFactory.Identifier("obj"))
+                .WithType(SyntaxFactory.ParseTypeName("object"));
+
+            var ifStatement =
+                SyntaxFactory.IfStatement(SyntaxFactory.ParseExpression($"obj == null || typeof({identifier}) != obj.GetType()"),
+                    SyntaxFactory.Block(SyntaxFactory.ParseStatement("return false;")));
+
+            var castStatement = SyntaxFactory.ParseStatement(
+                    $"var value = ({identifier}) obj;{Environment.NewLine}");
```

Comment:
Why does this need an explicit newline? Doesn't it get laid out properly otherwise?